File: src/main/java/org/bundleproject/discover/utils/Log.java
package org.bundleproject.discover.utils;

public class Log {

    private static int indentCount = 0;

    public static void startSection(String message) {
        sendSectionInfo(message);
        indentCount++;
    }

    public static void endSection() {
        indentCount = Math.max(0, indentCount - 1);
    }

    public static void endSection(String message) {
        sendSectionInfo(message);
        endSection();
    }

    public static void sendSectionInfo(String message) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < indentCount; i++) {
            sb.append("  ");
        }
        info(sb.append(message).toString());
    }

    public static void resetSections() {
        indentCount = 0;
    }

    public static void info(String msg) {
        System.out.println("INFO: " + msg);
    }

    public static void err(String msg) {
        System.out.println("ERROR: " + msg);
    }

    public static void warn(String msg) {
        System.out.println("WARN: " + msg);
    }

}

File: src/main/kotlin/org/bundleproject/discover/utils/Log.kt
package org.bundleproject.discover.utils

object Log {
    private var indentCount = 0
    fun startSection(message: String?) {
        sendSectionInfo(message)
        indentCount++
    }

    fun endSection() {
        indentCount = 0.coerceAtLeast(indentCount - 1)
    }

    fun endSection(message: String?) {
        sendSectionInfo(message)
        endSection()
    }

    fun sendSectionInfo(message: String?) {
        val sb = StringBuilder()
        for (i in 0 until indentCount) {
            sb.append("  ")
        }
        info(sb.append(message).toString())
    }

    fun resetSections() {
        indentCount = 0
    }

    fun info(msg: String) {
        println("INFO: $msg")
    }

    fun err(msg: String) {
        println("ERROR: $msg")
    }

    fun warn(msg: String) {
        println("WARN: $msg")
    }
}
